File: leetcode-solutions/ruby/350. Intersection of Two Arrays II.rb
# https://leetcode.com/problems/intersection-of-two-arrays-ii/

# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @return {Integer[]}
def intersect(nums1, nums2)
  map = {}
  ans = []
  nums1.each do |n|
      if map[n]
          map[n] += 1
      else
          map[n] = 1
      end
  end
  
  nums2.each do |n|
      if map[n] && map[n] > 0
         ans << n
          map[n] -= 1
      end
  end
  
  ans
end

File: leetcode-solutions/python/350. Intersection of Two Arrays II.py
# https://leetcode.com/problems/intersection-of-two-arrays-ii/
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums1.sort()
        nums2.sort()
        result = []
        i = 0
        j = 0
        while i < len(nums1) and j < len(nums2):
            if nums1[i] == nums2[j]:
                result.append(nums1[i])
                i += 1
                j += 1
            elif nums1[i] > nums2[j]:
                j += 1
            else:
                i += 1
        return result

# solution using dictionary/map
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        memo = defaultdict(int)
        for num in nums1:
            memo[num] += 1
            
        ans = []
        for num in nums2:
            if memo[num] > 0:
                ans.append(num)
                memo[num] -= 1
        return ans
